### PR TITLE
Improve click to focus

### PR DESCRIPTION
**Pull request by dkochmanski**

### CODE REVIEW

File: Core/clim-basic/ports.lisp [hunk 1]
```diff
@@ -84,28 +86,31 @@
...
-  ;; This was the case on Genera and is requested by users from time to time...
-  #+ (or)
-  (unless (eq *application-frame* (pane-frame focus))
-    (error "frame mismatch in ~S" '(setf port-keyboard-input-focus)))
-  (setf (port-frame-keyboard-input-focus port *application-frame*) focus))
-
-(defgeneric port-frame-keyboard-input-focus (port frame))
-(defgeneric (setf port-frame-keyboard-input-focus) (focus port frame))
+(defgeneric note-input-focus-changed (sheet state)
+  (:documentation "~
```
Comment (scymtym) [inline Core/clim-basic/ports.lisp]: The `~` seems wrong.

File: Core/clim-basic/ports.lisp [hunk 2]
```diff
@@ -347,17 +352,33 @@
              (map nil #'synthesize-enter sheets))
           (when (should-synthesize-for-sheet-p sheet)
             (push sheet sheets)))))
-
     new-pointer-sheet))
 
+(defmethod distribute-event ((port basic-port) (event keyboard-event))
+  ;; If no sheet is focused we do not dispatch the event.
+  (break "HI")
```
Comment (scymtym) [inline Core/clim-basic/ports.lisp]: Is this leftover from debugging?

File: Core/clim-basic/ports.lisp [hunk 3]
```diff
@@ -347,17 +352,33 @@
...
+  (when-let ((focused (port-keyboard-input-focus port)))
+    (dispatch-event-copy focused event)))
+
+;;; We focus sheet in CLIM thread.
+#+ (or)
+(defmethod distribute-event ((port basic-port) (event window-manager-focus-event))
```
Comment (scymtym) [inline Core/clim-basic/ports.lisp]: Can this be removed?

File: Core/clim-basic/stream-input.lisp [hunk 4]
```diff
@@ -179,7 +179,8 @@ keys read."))
     :documentation "Holds the last gesture returned by stream-read-gesture
 (not peek-p), untransformed, so it can easily be unread.")))
 
-(defmethod stream-set-input-focus ((stream standard-extended-input-stream))
+;; Deliberely not specialized to work on sheets too. -- jd 2019-08-23
```
Comment (scymtym) [inline Core/clim-basic/stream-input.lisp]: Deliberately

File: Core/clim-basic/ports.lisp [hunk 5]
```diff
@@ -385,13 +406,15 @@
...
-      ;; currently none.
+      ;; Pressing pointer button over a sheet makes a sheet pressed
+      ;; and focused. Pressed sheet is assigned only when there is
+      ;; currently none while focused event is always sent.
       (pointer-button-press-event
        (when (null pressed-sheet)
-         (setf (port-pressed-sheet port) new-pointer-sheet)))
-      ;; Releasing the button sets the pressed sheet to NIL.
+         (setf (port-pressed-sheet port) new-pointer-sheet))
+       (dispatch-event-copy new-pointer-sheet event 'window-manager-focus-event))
```
Comment (scymtym) [inline Core/clim-basic/ports.lisp]: I'm not sure about the "window manager" part in `window-manager-focus-event`. Isn't `window-manager-focus-event` for the case in which a frame gets the focus on the window manager/desktop environment level?

File: Core/clim-basic/ports.lisp (hunk 1 above)
Comment (dkochmanski) [reply Core/clim-basic/ports.lisp]: fixed

File: Core/clim-basic/ports.lisp (hunk 3 above)
Comment (dkochmanski) [reply Core/clim-basic/ports.lisp]: fixed

File: Core/clim-basic/stream-input.lisp (hunk 4 above)
Comment (dkochmanski) [reply Core/clim-basic/stream-input.lisp]: fixed

File: Core/clim-basic/ports.lisp (hunk 5 above)
Comment (dkochmanski) [reply Core/clim-basic/ports.lisp]: in principle the top-level-sheet receives this event when it is focused (on CLX; this event is not specified in CLIM II), In principle we could have enabled other mirrored sheets to receive such event (but we don't do that).It is used only to implement input focus, so I see no reason to inflate number of different events (like we do not introduce separate class for synthesized boundary events).

File: Core/clim-basic/ports.lisp (hunk 2 above)
Comment (dkochmanski) [reply Core/clim-basic/ports.lisp]: yes, fixed

**Comment by dkochmanski** [comment]
where do we stand with this?

**Comment by dkochmanski** [comment]
are we waiting with this branch for keyboard focus specificiation or should I rebase and resolve merge conflicts?

**Comment by dkochmanski** [comment]
rebased against master

**Comment by dkochmanski** [comment]
I've rebased this against master and converted to draft, before merging that I want to have refactored input streams in place.

**Comment by dkochmanski** [comment]
This branch is rebased on top of the master branch. It does not introduce any user-facing changes, but rather sorts out the input-focus behavior in McCLIM and enables other improvements (like input cycling).

**Pull request merged**